src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBwaAligner.java

```diff
@@ -0,0 +1,100 @@
+package org.broadinstitute.hellbender.tools.spark.pathseq;
+
+
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMFlag;
+import htsjdk.samtools.SAMRecord;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.bwa.*;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.SAMRecordToGATKReadAdapter;
+import scala.Tuple2;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Loads Bwa index and aligns reads. This is not a general Bwa aligner because it only retrieves primary alignments.
+ * If preserveAlignments is set, existing alignment data in each read is not overwritten, and any new alignments
+ * are added to the SA tag.
+ */
+public final class PSBwaAligner {
+
+    private final BwaMemIndex bwaIndex;
+    private final int minSeedLength, numThreads, maxAltHits, scoreThreshold;
+    private final boolean pairedAlignment;
+
+    public PSBwaAligner(final PSBwaArgumentCollection bwaArgs, final boolean pairedAlignment) {
+        this.bwaIndex = BwaMemIndexSingleton.getInstance(bwaArgs.bwaImage);
+        this.minSeedLength = bwaArgs.seedLength;
+        this.numThreads = bwaArgs.bwaThreads;
+        this.maxAltHits = bwaArgs.maxAlternateHits;
+        this.scoreThreshold = bwaArgs.scoreThreshold;
+        this.pairedAlignment = pairedAlignment;
+    }
+
+    public Iterator<GATKRead> apply(final Iterator<GATKRead> itr, final SAMFileHeader header) {
+        //Create aligner and set options
+        final BwaMemAligner aligner = new BwaMemAligner(bwaIndex);
+        if (pairedAlignment) {
+            aligner.alignPairs();
+        }
+        aligner.setMaxXAHitsAltOption(maxAltHits);
+        aligner.setMaxXAHitsOption(maxAltHits);
+        aligner.setMinSeedLengthOption(minSeedLength);
+        aligner.setOutputScoreThresholdOption(scoreThreshold);
+        aligner.setNThreadsOption(numThreads);
+
+        //Get list of reads on the partition
+        final List<GATKRead> reads = new ArrayList<>();
+        while (itr.hasNext()) {
+            reads.add(itr.next());
+        }
+        final int numReads = reads.size();
+        if (pairedAlignment && numReads % 2 != 0) {
+            throw new UserException.BadInput("Expected paired reads but there are an odd number");
+        }
+
+        //Align read sequences
+        final List<Tuple2<GATKRead, List<BwaMemAlignment>>> results;
```

Comment:
Rather than preparing this list, why don't you prepare your final output list directly by revising your input list.  Just refactor the code that starts on line 81 into a separate function that maps a GATKRead and list of alignments into a (possibly new) GATKRead.  On line 71 just modify the input list with the result of that call.